protocols/secio/src/exchange/impl_webcrypto.rs

```diff
@@ -22,115 +22,163 @@
+            .into_future()
+            .flatten()
+            .map(|key_pair| (key_pair, crypto))
+    });
+
+    // WebCrypto has generated a key-pair. Let's split this key pair into a private key and a
+    // public key.
+    let split_key = key_promise.and_then(move |(key_pair, crypto)| {
+        let private = js_sys::Reflect::get(&key_pair, &JsValue::from_str("privateKey"));
+        let public = js_sys::Reflect::get(&key_pair, &JsValue::from_str("publicKey"));
+        match (private, public) {
+            (Ok(pr), Ok(pu)) => Ok((pr, pu, crypto)),
+            (Err(err), _) => Err(err),
+            (_, Err(err)) => Err(err),
+        }
+    });
+
+    // Then we turn the public key into an `ArrayBuffer`.
+    let export_key = split_key.and_then(move |(private, public, crypto)| {
+        crypto.export_key("raw", &public.into())
+            .map(wasm_bindgen_futures::JsFuture::from)
+            .into_future()
+            .flatten()
+            .map(|public| ((private, crypto), public))
+    });
+
+    // And finally we convert this `ArrayBuffer` into a `Vec<u8>`.
+    let future = export_key
+        .map(|((private, crypto), public)| {
+            let public = js_sys::Uint8Array::new(&public);
+            let mut public_buf = vec![0; public.length() as usize];
+            public.copy_to(&mut public_buf);
+            (SendSyncHack(SendWrapper::new((private, crypto))), public_buf)
+        });
+
+    SendSyncHack(SendWrapper::new(future.map_err(|err| {
+        SecioError::IoError(io::Error::new(io::ErrorKind::Other, format!("{:?}", err)))
+    })))
 }
 
 /// Finish the agreement. On success, returns the shared key that both remote agreed upon.
 pub fn agree(algorithm: KeyAgreement, key: AgreementPrivateKey, other_public_key: &[u8], out_size: usize)
     -> impl Future<Item = Vec<u8>, Error = SecioError>
 {
-    let (tx, rx) = oneshot::channel();
-    let mut tx = Some(tx);
-
-    let curve = match algorithm {
-        KeyAgreement::EcdhP256 => "P-256",
-        KeyAgreement::EcdhP384 => "P-384",
-    };
-
-    let other_public_key = TypedArray::from(other_public_key).buffer();
-
-    let out_size = out_size as u32;
+    let (private_key, crypto) = key.0.take();
+
+    // We start by importing the remote's public key into the WebCrypto world.
+    let import_promise = {
+        let other_public_key = {
+            // This unsafe is here because the lifetime of `other_public_key` must not outlive the
+            // `tmp_view`. This is guaranteed by the fact that we clone this array right below.
+            let tmp_view = unsafe { js_sys::Uint8Array::view(other_public_key) };
+            js_sys::Uint8Array::new(tmp_view.as_ref())
+        };
 
-    let send = move |out: Reference| {
-        let _ = tx.take()
-            .expect("JavaScript promise has been resolved twice")       // TODO: prove
-            .send(out);
+        // Note: contrary to what one might think, we shouldn't add the "deriveBits" usage.
```

Comment:
Could you explain that bit?

Reply:
The `importKey` function takes a list of "usages": https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/importKey
For me it was obvious that one should pass the "deriveBits" usage, but if I remember correctly in practice both Firefox and Chrome produce an error if I do that.